metaworld/__init__.py

```diff
@@ -0,0 +1,185 @@
+"""Proposal for a simple, understandable MetaWorld API."""
+import abc
+import pickle
+from collections import OrderedDict
+from typing import List, NamedTuple, Type
+
+import metaworld.envs.mujoco.env_dict as _env_dict
+
+
+EnvID = str
+
+
+class Task(NamedTuple):
+    """All data necessary to describe a single MDP.
+
+    Should be passed into a MetaWorldEnv's set_task method.
+    """
+
+    env_id: EnvID
+    data: bytes  # Contains env parameters like random_init and *a* goal
+
+
+class MetaWorldEnv:
+    """Environment that requires a task before use.
+
+    Takes no arguments to its constructor, and raises an exception if used
+    before `set_task` is called.
+    """
+
+    def set_task(self, task: Task) -> None:
+        """Set the task.
+
+        Raises:
+            ValueError: If task.env_id is different from the current task.
+
+        """
+
+
+class Benchmark(abc.ABC):
+    """A Benchmark.
+
+    When used to evaluate an algorithm, only a single instance should be used.
+    """
+
+    @abc.abstractmethod
+    def __init__(self):
+        pass
+
+    @property
+    def train_classes(self) -> 'OrderedDict[EnvID, Type]':
+        """Get all of the environment classes used for training."""
+        return self._train_classes
+
+    @property
+    def test_classes(self) -> 'OrderedDict[EnvID, Type]':
+        """Get all of the environment classes used for testing."""
+        return self._test_classes
+
+    @property
+    def train_tasks(self) -> List[Task]:
+        """Get all of the training tasks for this benchmark."""
+        return self._train_tasks
+
+    @property
+    def test_tasks(self) -> List[Task]:
+        """Get all of the test tasks for this benchmark."""
+        return self._test_tasks
+
+
+_ML_OVERRIDE = dict(obs_type='plain', random_init=True)
+_MT_OVERRIDE = dict(obs_type='with_goal_id', random_init=False)
+
+_N_GOALS = 50
+
+
+def _encode_task(env_id, data):
+    return Task(env_id=env_id, data=pickle.dumps(data))
+
+
+def _make_tasks(classes, args_kwargs, kwargs_override):
+    tasks = []
+    for (env_id, args) in args_kwargs.items():
+        assert len(args['args']) == 0
+        env_cls = classes[env_id]
+        goals = [env_cls._sample_goal_do_not_use() for _ in range(_N_GOALS)]
+        for goal in goals:
+            kwargs = args['kwargs'].copy()
+            del kwargs['task_id']
+            kwargs.update(dict(goal=goal, env_cls=env_cls))
+            kwargs.update(kwargs_override)
+            tasks.append(_encode_task(env_id, kwargs))
+    return tasks
+
+
+def _ml1_env_names():
+    key_train = _env_dict.HARD_MODE_ARGS_KWARGS['train']
+    key_test = _env_dict.HARD_MODE_ARGS_KWARGS['test']
+    tasks = sum([list(key_train)], list(key_test))
+    assert len(tasks) == 50
+    return tasks
+
+
+class ML1(Benchmark):
```

Comment:
does this break the current import paths `from metaworld.benchmarks import ML1` etc?

there is code in the wild using paths, so if we are going to make a breaking change, we should gracefully report an error message and give instructions when someone tries to use the old API.

Reply:
The object doesn't implement the same interface anyways, so it wouldn't be compatible even if the path were the same.